storage/generic.go

```diff
@@ -0,0 +1,126 @@
+// Copyright 2020 The Prometheus Authors
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package storage
+
+// This file holds boilerplate adapters for generic MergeSeriesSet and MergeQuerier functions, so we can have one optimized
+// solution that works for both ChunkSeriesSet as well as SeriesSet.
+
+import "github.com/prometheus/prometheus/pkg/labels"
+
+type genericQuerierAdapter struct {
+	baseQuerier
+
+	// One-of. If both are set, Querier will be used.
+	q  Querier
+	cq ChunkQuerier
+}
+
+type genericSeriesSetAdapter struct {
+	SeriesSet
+}
+
+func (a *genericSeriesSetAdapter) At() Labeled {
+	return a.SeriesSet.At().(Labeled)
+}
+
+type genericChunkSeriesSetAdapter struct {
+	ChunkSeriesSet
+}
+
+func (a *genericChunkSeriesSetAdapter) At() Labeled {
+	return a.ChunkSeriesSet.At().(Labeled)
+}
+
+func (q *genericQuerierAdapter) Select(sortSeries bool, hints *SelectHints, matchers ...*labels.Matcher) (genericSeriesSet, Warnings, error) {
+	if q.q != nil {
+		s, w, err := q.q.Select(sortSeries, hints, matchers...)
+		return &genericSeriesSetAdapter{s}, w, err
+	}
+	s, w, err := q.cq.Select(sortSeries, hints, matchers...)
+	return &genericChunkSeriesSetAdapter{s}, w, err
+}
+
+func newGenericQuerierFrom(q Querier) genericQuerier {
+	if q == nil {
+		return nil
+	}
+	return &genericQuerierAdapter{baseQuerier: q, q: q}
+}
+
+func newGenericQuerierFromChunk(cq ChunkQuerier) genericQuerier {
+	if cq == nil {
+		return nil
+	}
+	return &genericQuerierAdapter{baseQuerier: cq, cq: cq}
+}
+
+type querierAdapter struct {
+	genericQuerier
+}
+
+type seriesSetAdapter struct {
+	genericSeriesSet
```

Comment:
It's a bit odd for this to be depending on something defined in fanout.go

Reply:
Moved here.